src/main/java/net/neoforged/neoforge/transfer/resources/IResourceStack.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.resources;
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.MapCodec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.function.BiFunction;
+import java.util.function.UnaryOperator;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.world.flag.FeatureElement;
+import net.minecraft.world.flag.FeatureFlagSet;
+import net.neoforged.neoforge.common.util.NeoForgeExtraCodecs;
+
+/**
+ * Represents the underlying instruction set for mutable and immutable resource stacks.
+ * This is provided as a helper and not necessary for a {@link net.neoforged.neoforge.transfer.handlers.resources.IResourceHandler IResourceHandler} to function.
+ *
+ * @param <T> resource type
+ */
+public interface IResourceStack<T extends IResource> {
```

Comment:
I am really not sure that these stack abstractions really make sense. In a lot of cases, working directly with `FluidStack`s and `ItemStack`s would probably be nicer. There needs to be a bigger discussion about this, so I am only leaving this one comment here for now.

Reply:
I agree with tech, and I have yet to see a good argument for why certain things need the mutability. As the implementations I have seen, either don't actually make use of it being mutable, or cause a lot of excess object churn by turning a bunch of things into mutable and back, instead of only changing them as needed